pooltool/evolution/event_based/simulate.py

```diff
@@ -223,26 +237,16 @@ def get_next_event(
     if collision_cache is None:
         collision_cache = CollisionCache.create()
 
+    transition_event = transition_cache.get_next()
+    if transition_event.time < event.time:
+        event = transition_event
+
     ball_table_event = get_next_ball_table_collision(
         shot, collision_cache=collision_cache
     )
-    if ball_table_event.time == shot.t:
-        # Ball-table collisions involving non-airborne balls always happen at dt=0 from
-        # the current simulation time and therefore in such a case we immediately return
-        # with the next event in hand. Not only does this speed up the algorithm (by
-        # avoiding calculation of any other possible events--none of which occur sooner
-        # than dt=0), but some of the code for other event predictions assume zero
-        # z-velocity and can trigger assertion errors if this condition is not met.
-        # Therefore, it is essential that this code block runs before any other event
-        # prediction (or at least any event prediction capable of producing event dt>0).
-        return ball_table_event
-    elif ball_table_event.time < event.time:
+    if ball_table_event.time < event.time:
```

Comment:
* Redefining ball states such that non-zero z-velocity balls are airborne solves the issue of event predictions assuming zero z-velocity
* We can probably still use these t=0 shortcuts though